platform-sdk/docs/proposals/noAccidentalBranches.md

```diff
@@ -0,0 +1,100 @@
+in our preconsensus event stream. Currently, the PcesWriter will discard ancient events without writing them to a PCES
+file. If that happens to a self event, we don't want to gossip it to any of our peers. Such an event is guaranteed to
+be a stale event, but without the window manager it is plausible to construct a scenario where a node forgets about a
+self event and ends up branching accidentally. The window manager enables us to filter out any such events before
+they can be gossiped.
+
+## SelfEventBuffer
+
+The "SelfEventBuffer" is a new component that is responsible for holding self events until those events become durable.
+
+The self event buffer receives as input a stream of all events coming out of the PcesWindowManager, including events
+created by other nodes. Its output is also a stream of events. When it receives an event created by another node,
+it immediately passes that through to the output. When it receives a self event, it stores that in an internal buffer
+until that event is durable in the PCES.
+
+Also as input, the self event buffer receives updates from the PcesWriter on the current durability status. It uses
+this to decide when/how to let self events through the buffer.
+
+A special edge case -- the self event buffer should not hold onto self events received through gossip. Such a thing
+is only plausible if a node's PCES files are either corrupted or lost, or after a very bizarre reconnect. By holding
+onto self events that were not received through gossip, we never disrupt the topological ordering of events (since
+it's impossible for other nodes to have built upon those events). But if we hold onto a self event that was received
+through gossip, we may inadvertently send events to gossip out of topological order. Additionally, if somehow don't
+know about a self event, holding that event in the self event buffer will not help us to avoid branching.
+
+As a debugging feature, we should log a rate limited error message if the self event buffer comes in contact with
+a self event that was received through gossip. Such an incident need not crash the node, but it is something that
+will possibly deserve investigation.
+
+## Feature Flag
+
+Until we decide we want to always leave this feature enabled, it should sit behind a feature flag. If not enabled,
+self events should reach gossip using the same pathway they do today.
+
+# Risks
+
+This new component risks increasing the time it takes for an event to reach consensus, since we are adding a delay
+between the creation of a self event and the time when the event becomes eligible to be gossiped.
+
+# Future Work
+
+An idea has been pitched where we make the PcesWriter super, super fast (possibly by memory-mapped files). In such a
+scenario, we might be able to remove the self event buffer entirely, and feed gossip from the events that come
+out of the writer.
```

Comment:
With this future in mind, why not simplify the design and nuke the `SelfEventBuffer` altogether? And I suppose the `PcesWindowManager` could be nuked too, although I'm not entirely sure.

The idea is to feed the `Gossip` from both the `PcesSqequencer` and the `PcesWriter` directly. The first input drops self events in the `Gossip` component because we only want to gossip them once they're persisted. The second input drops non-self events because they've already been gossiped through the first input.

Wouldn't this solve the problem statement with minimum changes to the existing architecture, or there's any flaws that the proposed more complex solution addresses?

Reply:
The main problem here is that we aren't 100% sure that it's possible to take the alternate approach, and the work to test its plausibility might end up being quite large.

Reply:
From what I see in this design PR and the related GitHub ticket, the proposed design hasn't been tested for its plausibility yet, too. If that is true, then I'd propose to start the test with the simplest design possible - e.g. the one I described above - and only introduce more complexity in the form of new components/wiring if the simple idea doesn't work.

I won't block your design if you strongly believe that it's superior to the simple solution. However, I'd really like to hear what other voters think, and ultimately I'd like to see at least some theoretical arguments that would show the advantage of the more complex solution and/or any flaws of my simple idea that I may have missed.